proto/sql_api.proto

```diff
@@ -0,0 +1,103 @@
+
+import "cockroach/proto/api.proto";
+import "cockroach/proto/data.proto";
+import "cockroach/proto/errors.proto";
+import "gogoproto/gogo.proto";
+
+option (gogoproto.sizer_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+// CmdRequestHeader is supplied with every CmdRequest.
+message CmdRequestHeader {
+  // Session settings that were returned in the last response that
+  // contained them, being reflected back to the server.
+  optional bytes session = 1;
+  // The transaction state returned in the previous response being
+  // reflected back.
+  optional bytes txn = 2;
+  // CmdID is optionally specified for request idempotence
+  // (i.e. replay protection).
+  optional ClientCmdID cmd_id = 3 [(gogoproto.nullable) = false, (gogoproto.customname) = "CmdID"];
+  // User is the originating user. Used to lookup priority when
+  // scheduling queued operations at target node.
+  optional string user = 4 [(gogoproto.nullable) = false];
+}
+
+// CmdResponseHeader is returned with every Cmd response.
+message CmdResponseHeader {
+  // Error is non-nil if an error occurred.
+  optional Error error = 1;
+  // Setting that should be reflected back in all subsequent requests.
+  // When not set, future requests should continue to use existing settings.
+  optional bytes settings = 2;
+  // Transaction message returned in a response; to be not interpreted by
+  // the sender and reflected in a subsequent request. When not set,
+  // the subsequent request should not contain a transaction object.
+  optional bytes txn = 3;
+}
+
+message Datum {
+  option (gogoproto.onlyone) = true;
+  oneof value {
+    bool bval = 1;
+    int64 ival = 2;
+    double dval = 3;
+    string sval = 4;
+    bytes blobval = 5;
+  }
+  // Checksum is a CRC-32-IEEE checksum of the value.
+  // If this is an integer value, then the value is interpreted as an 8
+  // byte, big-endian encoded value. This value is set by the client on
+  // updates to do end-to-end integrity verification. If the checksum is
+  // incorrect, the update operation will fail. If the client does not
+  // wish to use end-to-end checksumming, this value should be nil.
+  optional fixed32 checksum = 9;
+  // Timestamp of value in a result.
+  optional Timestamp timestamp = 10;
+}
+
+// A Result is a collection of values representing a row
+// in a result view. A column value not present in a row
+// has Nil Bytes in the value.
+message Result {
+  repeated Datum values = 1;
+}
+
+// A sql CmdRequest to cockroach. A transaction can consists of multiple
+// CmdRequests.
+message CmdRequest {
+  // Request header.
+  optional CmdRequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+    // SQL commands/queries to be serially executed by the server.
+  repeated string sql = 2;
+  // parameters are referred to in the above sql query using $1, $2, etc.
+  // params[n-1] -> $n
+  repeated Datum params = 3;
```

Comment:
The best reason I can think of is that it makes the mapping of command to ResultSet more complicated; you may pass what you think is one statement but get two ResultSets back. If we weren't already designing with batching in mind this would be more of a problem.

It looks like mysql eventually added the ability to opt-in to sending multiple statements; it's opt in in part because it's a change in the protocol (and client libraries) to send multiple result sets in response to one query.